### ISSUE TITLE
Different way of ending the program with GUI

### ISSUE THREAD

**Issue by wz2k** [nus-cs2103-AY2223S2/forum#99]
Is it possible to not have the 'bye' command after switching to GUI? With a GUI, users can simply close the program without typing. Also, when the program closes, users will not be able to read the bot's reply to the bye command making it even more redundant.

**Comment by 9fc70c892** [comment]
It is possible but you need to consider how to save if you have your Duke to save upon exit.

But what other ways are there? Clicking on top right/left hand side? Typing a different command? Alt F4?

Regarding second point 

You can get around the user not being able to read bye message by introducing some delay perhaps.

**Comment by 9fc70c892** [comment]
Now if your question is "is it against grading criteria", then sorry I am not able to answer your question (I'm not TA). But there is a need for safe exit if you wish to write tasks to file

**Comment by wz2k** [comment]
Thanks for the replies. I have handled the saving but I am concerned on the part of it being in the grading criteria.

**Comment by damithc** [comment]
@wz2k You are free to decide what's the best way to terminate the program. It's fine to replace/complement the 'bye' command with other mechanisms. A good topic for discussion.

> It is possible but you need to consider how to save if you have your Duke to save upon exit.

@9fc70c892 Yup, but not an issue if the data is saved at every command that changes data.

**Comment by hingen** [comment]
> Also, when the program closes, users will not be able to read the bot's reply to the bye command making it even more redundant.

I implemented my `bye` command to request from the user an additional line of input before closing using `Platform.exit()`. I think this is also a useful way to show any error messages before closing should the program run into any issues it cannot remedy (e.g. unable to read/create save data file when the app launches).

**Comment by damithc** [comment]
> I implemented my `bye` command to request from the user an additional line of input before closing using `Platform.exit()`. I think this is also a useful way to show any error messages before closing should the program run into any issues it cannot remedy (e.g. unable to read/create save data file when the app launches).

Interesting. Will the user become unhappy that s/he has to perform two steps to close the application, and chose to just close the Window manually instead?

**Comment by hingen** [comment]
> Will the user become unhappy that s/he has to perform two steps to close the application, and chose to just close the Window manually instead?

I implemented it such that the second input can just be an empty line. My original thought was that if a user did not care about seeing the exit message, they could just press enter twice.

However, now that I think about it, perhaps that isn't the best approach as users likely wouldn't care whether there is an exit message. I'll likely remove the extra input for the `bye` command but still keep it around for when the app encounters an issue it cannot remedy as I think it's better for the user to see the error message before the app closes.

Thanks for pointing that out!

**Comment by damithc** [comment]
> However, now that I think about it, perhaps that isn't the best approach as users likely wouldn't care whether there is an exit message. I'll likely remove the extra input for the `bye` command but still keep it around for when the app encounters an issue it cannot remedy as I think it's better for the user to see the error message before the app closes.

Good point. Incidentally, this brings up another downside of save-upon-exit strategy: i.e., the user may not have the time to deal with save-related errors when s/he is exiting the application. Imagine you are in a hurry to shut-down the computer to go somewhere and the application says tasks that you risk losing all tasks you entered earlier because there is an issue with saving!!!

**Comment by Tempura-Person** [comment]
in which case, it should be ok if my bye command does nothing, as long as Duke saves my tasklist properly after every command?

**Comment by damithc** [comment]
> in which case, it should be ok if my bye command does nothing, as long as Duke saves my tasklist properly after every command?

@Tempura-Person Yes, that's one option. The final decision is yours.

**Comment by eugenetangkj** [comment]
In the `handleUserInput` method, I checked if the input given by the user is "bye". If so, I made use of a `PauseTransition` object to close the stage after a specified duration. While the duration is still ongoing, the exit message will be processed and I prevented user input by using the `setDisable` method on `userInput` and `sendButton`. Then, after the duration, the application closes itself. Here's a useful StackOverflow [post](https://stackoverflow.com/questions/27334455/how-to-close-a-stage-after-a-certain-amount-of-time-javafx) that helped me.
